Tgmonkey

Trugath opened the pull request:
Robust Monkeys. Tested and in-use on BPL server. Only really a problem in hordes.

They steal things, aggro, disarm and beat you to crit, then stuff your body in disposals.
Keeping the changelog sparse so it's more Fun

🆑 Monkey Management
add: Added Monkey Behaviour
/🆑

Monkeys that step and make 4 random emotes are boring. Monkeys should be unpredictable, inquisitive and vindictive little bastards.

Trugath committed:
Monkeys are a bit more of a handful
They will aggro, steal, pick up weapons and stuff bodies in disposals

Trugath committed:
search using held_items instead of manually

Trugath committed:
Merge remote-tracking branch 'upstream/master' into tgmonkey

Lakart commented:
Genetics, Viro, and Xeno can be pretty stagnant. Having to beat a monkey to death every now and then adds spice to life.

And the kitchen is used to things that fight back when you try to make them into meat, or just because.

Arianya commented:
Does this mean I can't buckle a monkey to be my blood doll in Xeno anymore?

Arianya commented:
Also, looking forward to antag xenobiologist releasing a horde of monkeys

Trugath commented:
The resist out of buckle doesn't work based on my testing. they resist but the code doesn't pop them out the chair. decided that is a WIP. Didn't want to stomp over too much code. They do succesfully escape cuffs in about 90 seconds though.

PKPenguin321 commented:
oh boy oh boy oh boy!

Shadowlight213 commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,436 @@
+						  "[C] trips over [src] and falls!", \
+						  "[C] topples over [src]!", \
+						  "[C] leaps out of [src]'s way!")]</span>")
+		C.Weaken(2)
```
\>aggroed monkeys have the beepsky chasing stun.
Why?

Trugath replied:
It was hilarious. A horde of monkies will trip and dive over each other to get at you.

Trugath replied:
Feel free to remove, it wasn't in the original design. It just 'worked' nicely

Trugath replied:
I did add the check to stop them causing the stun when they are downed fully. So the window of Fun is smallish.

optimumtact commented:
@Cheridan

RemieRichards commented:
>then stuff your body in disposals

Absolutely not.

Cheridan commented:
@optimumtact Very yes

Trugath commented:
They will only try to disposals you if and only if: You attacked them directly or aggroed them in a way to be added to the Enemies list. You did so enough that them beating you didnt remove you from that list through the de-aggro mechanic and you are within sight of a disposals.
They do not press the handle.
This gives you time to just get out if you arn't in crit. If you are? there should be a cargo tech on duty to save you.

RemieRichards commented:
>They do not press the handle.

This is not a point in the favour of the mechanic, you do realise this right? being dumped in a disposal without it going of is essentially removing the person from the round forever, nobody checks disposal bins, you're just gone if it doesn't flush, there's an auto flush, but it's 100% obvious people will start turning that off when they know that they can ruin things for other people via monkeys.

Trugath commented:
Turn your suit sensors to Max. If it worries you.

optimumtact commented:
Don't disposals autoflush if things are in them?

Trugath commented:
The autoflush means them stuffing and not pressing the handle gives someone not in crit time to escape, someone in crit will be auto-flushed. RemieRcihards is worried about peaple disabling autoflush.

optimumtact commented:
? People already can crit other people and stuff them in disposals and that may or may not have autoflush turned off.

Seems like a silly point of order to complain about imo.

KorPhaeron commented:
This is great.

Hope it doesn't ruin the departments that use them though

Trugath commented:
Only thing i would really suggest is for Virology: Putting a mask on a monkey stops it biting. 
Stops unwanted infections.

PKPenguin321 commented:
maybe make domestic monkeys which are like what we already have to use for stuff like pun pun

Lakart commented:
They only hit you if you hit them, even a domesticated animal wont let you kick it around.

Shapsy commented:
I really doubt people would go around turning off autoflush on disposals just in case a monkey tried stuffing someone into one

Perhaps add a lobotomy surgery for monkies that removes all the behaviors this adds and makes them docile and unable to fight back, and make Pun Pun lobotimized by default? That'd probably be a lot of snowflake code though

AdamElTablawy commented:
so like goon monkeys then

fair warning-on goon, monkeys are extremely robust and routinely crit people who fuck with them.

similar things may happen here so make sure departments have straightjackets or something to avoid monkey RAGE destroying genetics

Lakart commented:
Not nearly as robust as goon monkeys, you could take 3 of them bare handed pretty easily.

AdamElTablawy commented:
lame

you haven't lived until you accidentally grab a monkey wrong on goon and 5 of them crash through the window suplex you onto a table and cut out your heart with a saw and scapel

Lakart commented:
5 of them will probably kill you.

MimicFaux commented:
I may have to give goon another chance. That sounds amazing.

MMMiracles commented:
saying no to monkey rage is saying no to fun

Armhulen commented:
@MimicFaux There's a special role called test subject that makes you a monkey assistant roundstart with a custom name

and monkeys love you, you can disposal dive into genetics and start VIVA LA MONKEY

RandomMarine commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -213,6 +213,18 @@
+			Weaken(5)
+			add_logs(M, src, "tackled")
+			visible_message("<span class='danger'>[M] has tackled down [src]!</span>", \
+				"<span class='userdanger'>[M] has tackled down [src]!</span>")
```
We're going to see lings going monkey just for guaranteed tackles.

Trugath replied:
Left the disarm, the stun is less reliable for controlled monkies now.

Trugath committed:
Only natural monkies should reliably stun
Natural monkies don't try to stun-lock

FantasticFwoosh commented:
@Armhulen now all sentient monkeys need slime-esque follows & rallies (active ability for infected monkeys?) as well as a attack so you can gather up monkeys and send them en-masse at people kicking & biting.

Really actually you could swap the 'test subject' out with anything, id personally use a Gorilla for a bit of muscle and variety (and literally nobody is going to mess with a gorilla unlike a chef chasing the monkeys) but playing a monkey assistant sounds great.

jughu commented:
PM TO ADMINS:HELP A MONKEY JUST KILLED ME AND STUFFED MY BODY IN A DISPOSAL IS HE TRAITOR?
ADMIN PM TO PLAYER:no thats just how monkeys always act from now on
PM TO ADMINS:FUCKING CODERS!!!

jughu commented:
I just gonnaa open 15 monkeycubes in armoury i dare anyone to attack it

Trugath commented:
Aslong as the players don't attack any of them, they will more or less ignore the players. They will snatch things out of hands though.  They only aggro due to action.

jughu commented:
I tottaly dont see this used against nuke ops
MONKEYTIDE!!! 
and they shoot shoot shoot shoot them untill they run out of bullets

Trugath commented:
A horde of monkeys are a serious threat if you hit them accidentally.

jughu commented:
Do they aggro on everything or only the attacker?

Chocobro commented:
:+1:

Trugath commented:
Only the attacker.

jughu commented:
That sucks tbh ENRAGED MONKEYS should be hostile to everything for a while

Trugath commented:
I didn't want people to monkeycube, smack and run off. I want spice and flavourful monkeys. Not hate driven murder machines.  It would be fine if they didn't swarm. With the swarm, I worry any generalised murder mechanic could be abused.

AnturK commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,436 @@
+	..()
+
+/mob/living/carbon/monkey/proc/take_to_slot(obj/item/G)
+	var/list/slots = list ("left hand" = slot_hands,"right hand" = slot_hands)
```
What's the point of these helpers, if they try same thing twice.

AnturK commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,436 @@
+	equip_in_one_of_slots(G, slots)
+	// update_hands = 1
+
+/mob/living/carbon/monkey/proc/equip_in_one_of_slots(obj/item/I, list/slots, qdel_on_fail = 0)
```
I'm pretty sure we have this on humans already, just move it to /carbon

AnturK reviewed (commented):
Just quick look, will read over it later.

Trugath committed:
Pushes equip_in_one_of_slots to carbon from human

Trugath committed:
Removed unused helper.

Trugath committed:
Make the trip and stun only against monkeys
The main reason this is here is for the swarm behaivour.

Trugath committed:
Just return true for now. It should flee on next life tick.

coiax commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,436 @@
+			if(health < 50)
+				mode = MONKEY_FLEE
+				spawn(1)
+					handle_combat()
```
SPAWNS ARE EVIL, USE addtimer

Trugath committed:
remove blocking call from Life

Trugath committed:
stuff_mob_in blocks so make it non-blocking

Trugath commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,436 @@
+// handle de-aggro
+/mob/living/carbon/monkey/proc/attacked(mob/living/carbon/H)
+	if(prob(25))
+		enemies[H] --
```
Incrementing uninitilised member of a list sets it to 1. I'm using the list as a Map. Mapping from Mob to Aggression. If a mob is in the map it is an enemy. The number is used in the de-aggro mechanic.

Trugath replied:
In the wierd case of the mob not being in enemies but attacked by the monkey, it gets added to enemies at -1 then immedietly removed. I could add a check... Okay i'll add the check

Trugath committed:
add check to stop add/remove in same function
Only occours if the monkey attacks someone who isn't in the enemies list.
This only happens during a monkey swarm. Best to check.

Trugath committed:
Fix for monkeys ripping stuff off walls.

Anonmare commented:
I'm sure someone could add a rage chemical later on. After that, just pass your monkey a dual e-sword, enrage them and run the hell away.

Militaires commented:
@mimicfaux
come over
i suggest LLJK 1 if you like RP, 2 if otherwise, goon has some pretty fun stuff.

Trugath commented:
They snatch from your hands. They don't take from the pocket slots.

Saegrimr commented:
dicks out

Trugath committed:
Moved shared emote/weaken down into living
Called the new function knockOVer. Monkey and Secbot both have different
control flow before the shared code.

Trugath commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,440 @@
+			retaliate(H)
+	..()
+
+/mob/living/carbon/monkey/Crossed(atom/movable/AM)
```
Struggled with this, Was copied originally but the control flow is different between the two mobs. Best I could think of without risking breaking something was pushing the emote and weaken down into a proc in living. I called the proc knockDown but i appreciate suggestions on improving this.

Trugath commented:
There is a lot of code shared between this, secbot and ed209. The code isn't easily de-duplicated because it relies on state only in these mobs. Additionally the goals and concepts of the patch (AI behaivour) are in other places in the code as well.

I guess the one long term solution is to rafactor everything into snpc/interactive and make that more general for living things as opposed to just humans?
The another is to do something with brains. Put the behaivour in them and let people switch brains around. This would let someone code the lobotomy surgery that was suggested earlier.

There is a world of options, but really I just want less boring lifeless monkeys.

jughu commented:
The people who downvoted this are unrobust cucks

AdamElTablawy commented:
@Trugath 
tbh i miss the days of giving monkeys an esword, punching them and RUNNING and watching the monkey murder anything that came near it on goon, so them being aggro on everything would be p funny

@MimicFaux 

you haven't lived until you see a wizard get his shit kicked in by four monkeys with knives

no seriously there was a wizard one round and he accidentally punched a monkey and like five of them ran out of genetics suplexed him to the floor and started stabbing him to death

Trugath committed:
Stop monkeys fighting over body

Trugath commented:
@AdamElTablawy If this get merged and people don't revolt, we can think about possible changes.

AnturK commented:
How WIP is this ? I suppose testmerge would be in order.

Trugath commented:
Apart from any reported bugs, balance requests and suggested tweaks? I'm done for now.

AnturK commented:
Judging from testmerge, works fine. 
Could probably use a tune up in strength and some aggro on special effects (flashes and such) but that's secondary.

AnturK commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,433 @@
+
+
+/mob/living/carbon/monkey
+	var/aggressive=0
```
This variable is not used ?

AnturK replied:
Still not used. Mostly want this implemented so admin created monkey hordes are dangerous. But removal also works.

dannno commented:
Yeah honestly they aren't deadly enough
I wanted to punch a genetics monkey and be literally torn apart
also weren't they supposed to be friendly if you are also a monkey? because they fucked me up regardless

MrStonedOne commented:
I'd prefer they came in too weak, then got buffed later on after some play time, then the reverse.

Iamgoofball commented:
What if they used deep learning

optimumtact commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,433 @@
+	else if(istype(I,/obj/item/clothing))
+		var/obj/item/clothing/C = I
+		monkeyDrop(C)
+		spawn(5)
```
why do you have spawns here

Trugath replied:
These spawns are a non-blocking delay. I can look into replacing them with something else.

optimumtact commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,433 @@
+				if(!pickpocketing)
+					pickpocketing = TRUE
+					M.visible_message("[src] starts trying to take [pickupTarget] from [M]", "[src] tries to take [pickupTarget]!")
+					spawn(5)
```
another spawn?

Trugath replied:
do_mob is a blocking call, the spawn removes the blocking call from life.

optimumtact commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,433 @@
+			retaliate(pulledby)
+			return TRUE
+
+	// nrrr, RARRRGH
```
please dont' leave in commented out code

optimumtact commented on code/modules/mob/living/carbon/monkey/combat.dm:
```diff
@@ -0,0 +1,433 @@
+/mob/living/carbon/monkey/proc/monkeyDrop(var/obj/item/A)
+	if(A)
+		unEquip(A)
+		A.loc = get_turf(src) // drop item works inconsistently
```
This should never happen, and even if it does you should use A.forceMove(get_turf(src))

optimumtact replied:
Also some items can get deleted on drop so you should check !qdeleted(A) after

Trugath committed:
Remove commented out code

Trugath committed:
Reduce flee range

Trugath committed:
Merge remote-tracking branch 'upstream/master' into tgmonkey

optimumtact commented:
This has been sitting in review hell because personally I'm lukewarm on the pr so I don't want to review and merge it myself, and all the other maintainers are basically lazy as shit.

Sorry 😢

RemieRichards commented:
I'm not lazy, I just hate the concept, and the code of it I've seen is just copypasted SNPC code.
Not worth merging tbh.

Trugath commented:
It's okay. I never expected it to get merged. Someone bugged me to submit it it as a PR. 

My original plan was to isolate the new behaviour in minimal code changes to the existing code-base. Hence the 'style' of code. I got the desired behaviour working using accepted code from the codebase as templates. The main part of the code is the state machine, everything around it is supporting code and is prime target for refactoring.

Rewriting it to obscure the fact I did template it from interactive and secbot would have just gone against any refactoring effort later.

I tried dropping into the irc channel to discuss the PR and it's code a few times. I didn't get any response, which is fine, i understand it might have been bad timing. I was sadly asleep (I'm on GMT) both times I know it was test merged.

This was a tentative 'Trying not to step on maintainers toes' PR. I can be more aggressive with the code if required.

Trugath committed:
Merge remote-tracking branch 'upstream/master' into tgmonkey

RemieRichards commented:
I actually wrote a review just now but misclicked and closed chrome (fuuuuu), the gist of it is this:
* ```spawn()``` -> ```addtimer()``` (look into this proc)
* Defines/Consts for magic values
* ```locate()```s are bad so try to store them in vars instead of redoing them (there's atleast 2-3 cases in here I've seen)

Trugath committed:
replace spawn with addTimer by moving code into callbacks

Trugath committed:
Move magic numbers to defines
Also removed 900 delay in resisting cuffs as its a useless argument

Trugath committed:
Refactor. Attacking has a function: monkey_attack
Combined de-aggro into it

RemieRichards commented:
Much nicer.

Trugath committed:
rework monkeyDrop to force drop
Also check for deleted items when snatching items

Trugath committed:
Aggresive can be toggled by an admin
Allows for a mass mob of angry monkies

Trugath committed:
Tweak aggressive monkeys to attack each other less
Monkey vs Monkey probability is a define
Warning: Once MVM combat breaks out in large mobs, mayhem insues

Trugath committed:
Set the monkey back to idle once it attemps to dispose the body

Trugath committed:
fix for aggressive monkeys self-harming

Trugath committed:
comment wording fix

Trugath committed:
changed walk2derpless to a callback and fixed deathgrip

AnturK commented:
@Trugath Is it ready ? The major issues are gone, but don't know if you're done with the changes.

Trugath commented:
It's done. I'll hands off till someone complains about something.

Cyberboss commented:
Inb4 geneticists/viros valid all monkeys
